torch/csrc/distributed/c10d/ProcessGroup.cpp

```diff
@@ -158,3 +159,105 @@ void ProcessGroup::release_resources() {
 }
 
 } // namespace c10d
+
+namespace {
+
+class WorkRegistry {
+ public:
+  void register_work(
+      const at::Tensor& tensor,
+      const c10::intrusive_ptr<c10d::Work>& work) {
+    auto storage = tensor.storage().getWeakStorageImpl();
+    std::unique_lock lock(lock_);
+    auto [it, inserted] = registry_.try_emplace(std::move(storage), work);
+    TORCH_CHECK(
+        inserted || it->second != work,
+        "The tensor storage is already associated with another work.");
+  }
+
+  c10::intrusive_ptr<c10d::Work> pop_work(const at::Tensor& tensor) {
+    const auto storage = tensor.storage().getWeakStorageImpl();
+    std::unique_lock lock(lock_);
+    auto it = registry_.find(storage);
+    if (it == registry_.end()) {
+      return nullptr;
+    }
+    auto work = it->second;
+    registry_.erase(it);
+    return work;
+  }
+
+  void unregister_completed_works() {
+    std::unique_lock lock(lock_);
+    for (auto it = registry_.begin(); it != registry_.end();) {
+      if (it->second->isCompleted()) {
+        it = registry_.erase(it);
+      } else {
+        ++it;
+      }
+    }
+  }
+
+  size_t get_work_registry_size() {
+    std::unique_lock lock(lock_);
+    return registry_.size();
+  }
+
+  ~WorkRegistry() {
+    unregister_completed_works();
```

Comment:
Unfortunately there is no reliable callback method where we can insert the "unregister_work()" cleanup call (`Work::finish()` is only called in Gloo PG but not NCCL PG, thus we can't use it). Instead we can do `unregister_completed_works()` here to do the cleanup for completed works before throwing the warning for uncompleted work.